src/daft-writers/src/file.rs

```diff
@@ -52,53 +86,59 @@ impl FileWriter for TargetFileSizeWriter {
         );
         use std::cmp::Ordering;
+        if input.is_empty() {
+            return Ok(());
+        }
 
+        let input_size_bytes = input.size_bytes()?.expect(
+            "Micropartitions should be loaded before writing, so they should have a size in bytes",
+        );
+        let avg_row_size_bytes = max(input_size_bytes / input.len(), 1);
+        // Write the input, rotating the writer when the current file reaches the target size
         let mut local_offset = 0;
-
         loop {
+            let rows_until_target = max(
+                self.remaining_bytes_for_current_file() / avg_row_size_bytes,
+                1,
+            );
             let remaining_input_rows = input.len() - local_offset;
-            let rows_until_target = self.target_in_memory_file_rows - self.current_file_rows;
-
             match remaining_input_rows.cmp(&rows_until_target) {
-                Ordering::Equal => {
-                    // Write exactly what's needed to fill the current file
-                    let to_write =
-                        input.slice(local_offset, local_offset + remaining_input_rows)?;
-                    self.current_writer.write(&to_write.into())?;
-                    self.rotate_writer()?;
-                    self.current_file_rows = 0;
-                    return Ok(());
-                }
-                Ordering::Less => {
-                    // Write remaining input and update counter
+                // We have enough rows to finish the file, write it, rotate the writer and return
+                Ordering::Equal | Ordering::Less => {
                     let to_write =
                         input.slice(local_offset, local_offset + remaining_input_rows)?;
-                    self.current_writer.write(&to_write.into())?;
-                    self.current_file_rows += remaining_input_rows;
+                    self.write_and_update_bytes(
+                        to_write.into(),
+                        remaining_input_rows * avg_row_size_bytes,
+                    )?;
                     return Ok(());
                 }
+                // We have more rows to write, write the target amount, rotate the writer and continue
                 Ordering::Greater => {
-                    // Write what fits in current file
                     let to_write = input.slice(local_offset, local_offset + rows_until_target)?;
-                    self.current_writer.write(&to_write.into())?;
-                    self.rotate_writer()?;
-                    self.current_file_rows = 0;
-
-                    // Update offset and continue loop
+                    self.write_and_update_bytes(
+                        to_write.into(),
+                        rows_until_target * avg_row_size_bytes,
+                    )?;
                     local_offset += rows_until_target;
                 }
             }
         }
     }
 
+    fn tell(&self) -> DaftResult<Option<usize>> {
```

Comment:
since we don't actually need `tell`, we could update this to `bytes_written` in the trait since that's what we really care about. Then it would be monotonically increasing. For the multiwriter case, it could just sum up across all the files.

Reply:
Makes sense, implemented